Source/Scene/ParticleSystem.js

```diff
@@ -0,0 +1,295 @@
+            this.maxWidth = defaultValue(options.maxWidth, 1.0);
+        }
+
+        var height = defaultValue(options.height, undefined);
+        if (height) {
+            this.minHeight = height;
+            this.maxHeight = height;
+        }
+        else {
+            this.minHeight = defaultValue(options.minHeight, 1.0);
+            this.maxHeight = defaultValue(options.maxHeight, 1.0);
+        }
+
+        this.lifeTime = defaultValue(options.lifeTime, Number.MAX_VALUE);
+
+        this.complete = new Event();
+        this.isComplete = false;
+
+        this.show = defaultValue(options.show, true);
+
+        this.carryOver = 0.0;
+        this.currentTime = 0.0;
+        this._billboardCollection = undefined;
+        this._previousTime = null;
+    };
+
+    function removeBillboard(system, particle) {
+        system._billboardCollection.remove(particle._billboard);
+    }
+
+    function random(a, b) {
+        return CesiumMath.nextRandomNumber() * (b - a) + a;
+    }
+
+    function updateBillboard(system, particle) {
+        var billboard = particle._billboard;
+        if (!defined(billboard)) {
+            billboard = particle._billboard = system._billboardCollection.add({
+                image: particle.image
+            });
+        }
+        billboard.width = particle.size.x;
+        billboard.height = particle.size.y;
+        billboard.position = particle.position;
+
+        // Update the color
+        var r = particle.startColor.red + particle.normalizedAge * (particle.endColor.red - particle.startColor.red);
+        var g = particle.startColor.green + particle.normalizedAge * (particle.endColor.green - particle.startColor.green);
+        var b = particle.startColor.blue + particle.normalizedAge * (particle.endColor.blue - particle.startColor.blue);
+        var a = particle.startColor.alpha + particle.normalizedAge * (particle.endColor.alpha - particle.startColor.alpha);
+        billboard.color = new Color(r,g,b,a);
+
+        // Update the scale
+        var scale = particle.startScale + particle.normalizedAge * (particle.endScale - particle.startScale);
+        billboard.scale = scale;
+    }
+
+    ParticleSystem.prototype.add = function(particle) {
+        particle.startColor = Color.clone(this.startColor);
+        particle.endColor = Color.clone(this.endColor);
+        particle.startScale = this.startScale;
+        particle.endScale = this.endScale;
+        particle.image = this.image;
+        particle.life = random(this.minLife, this.maxLife);
+        particle.mass = random(this.minMass, this.maxMass);
+
+        particle.size = new Cartesian2(random(this.minWidth, this.maxWidth), random(this.minHeight, this.maxHeight));
+
+        var speed = random(this.minSpeed, this.maxSpeed);
+        Cartesian3.multiplyByScalar(particle.velocity, speed, particle.velocity);
+
+        this.particles.push(particle);
+    };
+
+    ParticleSystem.prototype.calcNumberToEmit = function(dt) {
```

Comment:
I'll rename it, but also mark it as @private.  It's not something that an external user should be calling.